daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/SeparatedSequenceUnparsers.scala

```diff
@@ -0,0 +1,423 @@
+
+/**
+ * DFDL Spec. section 14.2.3 specifies only a few different behaviors
+ * for separator suppression. Each has an algorithm.
+ */
+sealed trait SeparatorSuppressionAlgorithm extends Serializable {
+
+  /**
+   * Determines if we should suppress a zero length element and
+   * separator.
+   *
+   * Checks if the length is zero if necessary to decide.
+   *
+   * Returns Suppress or IfTrailing when the element is, in fact, zero length,
+   * and the algorithm wants those to be suppressed.
+   * Returns DoNotSuppress otherwise.
+   */
+  def shouldSuppressIfZeroLength(state: UState, infosetElement: DIElement): SeparatorSuppressionAction
+
+  def shouldDoExtraSeparators = false
+
+  def checkArrayPos(state: UState, maxRepeats: Long): Boolean = true
+
+}
+
+object SeparatorSuppressionAlgorithm {
+  type Type = SeparatorSuppressionAlgorithm
+  import SeparatorSuppressionAction._
+
+  object None extends Type {
+    def shouldSuppressIfZeroLength(state: UState, infosetElement: DIElement) = Assert.usageError("not to be called.")
+  }
+  object FixedOrExpression extends Type {
+    def shouldSuppressIfZeroLength(state: UState, infosetElement: DIElement) = DoNotSuppress
+  }
+
+  class SuppressAnyEmpty(zeroLengthDetector: ZeroLengthDetector) extends Type {
+    def shouldSuppressIfZeroLength(state: UState, infosetElement: DIElement) = {
+      val isZL = zeroLengthDetector.isZeroLength(infosetElement)
+      if (isZL) Suppress
+      else DoNotSuppress
+    }
+  }
+
+  trait CheckArrayPosMixin { self: SeparatorSuppressionAlgorithm =>
+    override def checkArrayPos(state: UState, maxRepeats: Long): Boolean = {
+      state.arrayPos <= maxRepeats
+    }
+  }
+
+  class ImplicitSuppressAnyEmpty(zeroLengthDetector: ZeroLengthDetector)
+    extends SuppressAnyEmpty(zeroLengthDetector)
+    with CheckArrayPosMixin
+
+  object ImplicitNeverOrNotPotentiallyTrailing extends Type
+    with CheckArrayPosMixin {
+    def shouldSuppressIfZeroLength(state: UState, infosetElement: DIElement) = DoNotSuppress
+
+    override def shouldDoExtraSeparators = true
+
+  }
+
+  class ImplicitPotentiallyTrailing(zeroLengthDetector: ZeroLengthDetector) extends Type
+    with CheckArrayPosMixin {
+    def shouldSuppressIfZeroLength(state: UState, infosetElement: DIElement) = {
+      val isZL = zeroLengthDetector.isZeroLength(infosetElement)
+      if (isZL) IfTrailing
+      else DoNotSuppress
+    }
+  }
+}
+
+sealed trait SeparatorSuppressionAction extends Serializable
+object SeparatorSuppressionAction {
+  type Type = SeparatorSuppressionAction
+  case object Suppress extends Type
+  case object DoNotSuppress extends Type
+  case object IfTrailing extends Type
+}
+
+trait Separated { self: SequenceChildUnparser =>
+
+  def sep: Unparser
+  def spos: SeparatorPosition
+  def ssp: SeparatorSuppressionPolicy
+  def ssAlgorithm: SeparatorSuppressionAlgorithm
+
+  val childProcessors = Seq(childUnparser, sep)
+}
+
+class ScalarOrderedSeparatedSequenceChildUnparser(
```

Comment:
The noun is "SequenceChild" These are mods/flavors of SequenceChild unparsers. However SequenceChild is not a great name. Perhaps an abbreviation like SeqChild makes things easier? Or I could set it off with "_" though that's not our usual convention it may be worth it here.